File: app/control_plane/test/ankole/actor_runtime/worker_bootstrap_test.exs
defmodule Ankole.ActorRuntime.WorkerBootstrapTest do
  use Ankole.DataCase, async: false

  alias Ankole.ActorRuntime.WorkerAuthKey
  alias Ankole.ActorRuntime.WorkerBootstrap
  alias Ankole.ActorRuntime.WorkerBootstrap.Docker
  alias Ankole.ActorRuntime.WorkerBootstrap.Spec

  describe "container_spec/1" do
    test "describes the shared Agent Computer Docker runtime without inventing worker state" do
      assert {:ok,
              %Spec{
                contract_version: 2,
                kind: :container,
                image: "ankole-agent-computer:0.1.0",
                docker: %{
                  cap_add: ["SYS_ADMIN"],
                  security_opts: ["seccomp=unconfined", "systempaths=unconfined"],
                  extra_hosts: []
                },
                env: %{},
                host_setup_dirs: [],
                mounts: []
              }} =
               WorkerBootstrap.container_spec(image: "ankole-agent-computer:0.1.0")
    end
  end

  describe "worker_spec/1" do
    test "adds canonical auth, connectivity, and workspace guarantees" do
      workspace_root = "/tmp/ankole worker"

      assert {:ok,
              %Spec{
                contract_version: 2,
                kind: :worker,
                image: "ankole-agent-computer:0.1.0",
                docker: %{
                  cap_add: ["SYS_ADMIN"],
                  security_opts: ["seccomp=unconfined", "systempaths=unconfined"],
                  extra_hosts: [
                    %{host: "host.docker.internal", address: "host-gateway"}
                  ]
                },
                env: %{
                  "WORKER_ID" => "worker-a",
                  "RUNTIME_FABRIC_URL" => runtime_fabric_url
                },
                host_setup_dirs: [
                  "/tmp/ankole worker/shared/user-files",
                  "/tmp/ankole worker/shared/skills/agents",
                  "/tmp/ankole worker/sessions"
                ],
                mounts: [
                  %{
                    source: "/tmp/ankole worker/shared",
                    target: "/workspace/shared",
                    readonly: false
                  },
                  %{
                    source: "/tmp/ankole worker/sessions",
                    target: "/workspace/.sessions",
                    readonly: false
                  }
                ]
              } = spec} =
               WorkerBootstrap.worker_spec(
                 endpoint: "tcp://host.docker.internal:6010",
                 worker_id: "worker-a",
                 auth_key: "secret with / symbols",
                 image: "ankole-agent-computer:0.1.0",
                 workspace_root: workspace_root
               )

      assert runtime_fabric_url ==
               "tcp://:secret+with+%2F+symbols@host.docker.internal:6010"

      assert Docker.argv(spec) == [
               "run",
               "--rm",
               "--cap-add",
               "SYS_ADMIN",
               "--security-opt",
               "seccomp=unconfined",
               "--security-opt",
               "systempaths=unconfined",
               "--add-host",
               "host.docker.internal=host-gateway",
               "-e",
               "RUNTIME_FABRIC_URL=tcp://:secret+with+%2F+symbols@host.docker.internal:6010",
               "-e",
               "WORKER_ID=worker-a",
               "--mount",
               "type=bind,src=/tmp/ankole worker/shared,dst=/workspace/shared",
               "--mount",
               "type=bind,src=/tmp/ankole worker/sessions,dst=/workspace/.sessions",
               "ankole-agent-computer:0.1.0"
             ]

      shell = Docker.shell_command(spec)
      assert shell =~ "mkdir -p '/tmp/ankole worker/shared/user-files'"
      assert shell =~ "'type=bind,src=/tmp/ankole worker/shared,dst=/workspace/shared'"

      assert_raise ArgumentError, ~r/cannot replace canonical environment/, fn ->
        Docker.argv(spec, additional_env: %{"WORKER_ID" => "adapter-worker"})
      end
    end

    test "resolves the persisted AppConfigure auth key when no explicit key is passed" do
      assert {:ok, auth_key} = WorkerAuthKey.ensure()

      assert {:ok, %Spec{env: %{"RUNTIME_FABRIC_URL" => runtime_fabric_url}}} =
               WorkerBootstrap.worker_spec(
                 endpoint: "tcp://127.0.0.1:6010",
                 worker_id: "worker-app-config",
                 workspace_root: "/tmp/ankole-worker"
               )

      assert runtime_fabric_url ==
               "tcp://:#{URI.encode_www_form(auth_key)}@127.0.0.1:6010"
    end

    test "rejects invalid or incomplete worker inputs" do
      assert {:error, {:missing, :workspace_root}} =
               WorkerBootstrap.worker_spec(
                 endpoint: "tcp://127.0.0.1:6010",
                 worker_id: "worker-a",
                 auth_key: "secret"
               )

      assert {:error, {:invalid, :auth_key}} =
               WorkerBootstrap.worker_spec(
                 endpoint: "tcp://127.0.0.1:6010",
                 worker_id: "worker-a",
                 auth_key: "",
                 workspace_root: "/tmp/ankole-worker"
               )

      assert {:error, :invalid_runtime_fabric_endpoint} =
               WorkerBootstrap.worker_spec(
                 endpoint: "http://127.0.0.1:6010",
                 worker_id: "worker-a",
                 auth_key: "secret",
                 workspace_root: "/tmp/ankole-worker"
               )
    end
  end
end

File: tools/devkit/src/worker-bootstrap.test.ts
import { describe, expect, test } from 'bun:test'
import { readFileSync } from 'node:fs'
import { join } from 'node:path'
import { buildDockerRunArgs, parseWorkerBootstrapSpec, type WorkerBootstrapSpec } from './worker-bootstrap'
import { repoRootPath } from './utils'

const workerSpec: WorkerBootstrapSpec = {
  contract_version: 3,
  kind: 'worker',
  image: 'ankole-agent-computer:test',
  docker: {
    cap_add: ['SYS_ADMIN'],
    security_opts: ['seccomp=unconfined', 'systempaths=unconfined'],
    extra_hosts: [{ host: 'host.docker.internal', address: 'host-gateway' }]
  },
  env: {
    ANKOLE_AGENTS_ROOT: '/agents',
    ANKOLE_RUNTIME_FABRIC_ENDPOINT: 'tcp://host.docker.internal:6010',
    ANKOLE_RUNTIME_FABRIC_WORKER_AUTH_KEY: 'secret',
    WORKER_ID: 'worker-a'
  },
  host_setup_dirs: ['/tmp/ankole worker/agents'],
  mounts: [{ source: '/tmp/ankole worker/agents', target: '/agents', readonly: false }]
}

describe('parseWorkerBootstrapSpec', () => {
  test('parses the last complete v3 contract from Mix output', () => {
    expect(parseWorkerBootstrapSpec(`Compiling...\n${JSON.stringify(workerSpec)}\n`)).toEqual(workerSpec)
  })

  test('rejects the v2, unversioned, or incomplete contract', () => {
    expect(() => parseWorkerBootstrapSpec(JSON.stringify({ ...workerSpec, contract_version: 2 }))).toThrow(
      'unsupported worker bootstrap contract version'
    )
    expect(() => parseWorkerBootstrapSpec(JSON.stringify({ ...workerSpec, contract_version: undefined }))).toThrow(
      'unsupported worker bootstrap contract version'
    )
    expect(() => parseWorkerBootstrapSpec(JSON.stringify({ ...workerSpec, mounts: undefined }))).toThrow(
      'worker bootstrap contract is invalid'
    )
  })
})

describe('buildDockerRunArgs', () => {
  test('translates canonical fields and appends only caller-owned differences', () => {
    expect(
      buildDockerRunArgs(workerSpec, {
        name: 'ankole-dev-worker',
        labels: { 'ankole.dev.managed': 'true' },
        additionalMounts: [{ source: '/repo/src', target: '/repo/app/agent_computer/src', readonly: true }],
        command: ['/bin/sh', '-lc', 'exec bun --watch src/main.ts']
      })
    ).toEqual([
      'run',
      '--rm',
      '--name',
      'ankole-dev-worker',
      '--label',
      'ankole.dev.managed=true',
      '--cap-add',
      'SYS_ADMIN',
      '--security-opt',
      'seccomp=unconfined',
      '--security-opt',
      'systempaths=unconfined',
      '--add-host',
      'host.docker.internal=host-gateway',
      '-e',
      'ANKOLE_AGENTS_ROOT=/agents',
      '-e',
      'ANKOLE_RUNTIME_FABRIC_ENDPOINT=tcp://host.docker.internal:6010',
      '-e',
      'ANKOLE_RUNTIME_FABRIC_WORKER_AUTH_KEY=secret',
      '-e',
      'WORKER_ID=worker-a',
      '--mount',
      'type=bind,src=/tmp/ankole worker/agents,dst=/agents',
      '--mount',
      'type=bind,src=/repo/src,dst=/repo/app/agent_computer/src,readonly',
      'ankole-agent-computer:test',
      '/bin/sh',
      '-lc',
      'exec bun --watch src/main.ts'
    ])
  })
})

describe('production worker process topology', () => {
  test('keeps the image init as PID 1 instead of replacing it in Helm', () => {
    const dockerfile = readFileSync(join(repoRootPath, 'app', 'agent_computer', 'Dockerfile'), 'utf8')
    const deployment = readFileSync(
      join(repoRootPath, 'tools', 'deploy', 'helm', 'ankole-agent', 'templates', 'deployment.yaml'),
      'utf8'
    )
    const workerContainer = deployment.slice(deployment.indexOf('        - name: worker'))

    expect(dockerfile).toContain('ENTRYPOINT ["tini", "-g", "--"]')
    expect(workerContainer).not.toMatch(/^ {10}(command|args):/m)
    expect(workerContainer).toContain('name: ANKOLE_RUNTIME_FABRIC_ENDPOINT')
    expect(workerContainer).toContain('name: ANKOLE_RUNTIME_FABRIC_WORKER_AUTH_KEY')
  })
})
